cpp/src/internal/request_am.cpp

```diff
@@ -15,22 +15,30 @@ RecvAmMessage::RecvAmMessage(internal::AmData* amData,
                              ucp_ep_h ep,
                              std::shared_ptr<RequestAm> request,
                              std::shared_ptr<Buffer> buffer)
-  : _amData(amData), _ep(ep), _request(request), _buffer(buffer)
+  : _amData(amData), _ep(ep), _request(request)
 {
-  _request->_delayedSubmission =
-    std::make_shared<DelayedSubmission>(false, _buffer->data(), _buffer->getSize());
+  std::visit(data::dispatch{
+               [this, buffer](data::AmReceive& amReceive) { amReceive._buffer = buffer; },
+               [](auto arg) { throw std::runtime_error("Unreachable"); },
+             },
+             _request->_requestData);
```

Comment:
It _might_ (?) be cleaner in this case to write:
```
if (std::holds_alternative_v<data::AmReceive>(_request->_requestData)) {
    ...
} else {
   throw std::runtime_error("Unreachable!");
}
```
?

Reply:
No objections either way from me, but I'd prefer if we can keep things more or less consistent everywhere. See https://github.com/rapidsai/ucxx/pull/121#discussion_r1422430282.